tests/framework/matrix.py

```diff
@@ -0,0 +1,160 @@
+# Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+# SPDX-License-Identifier: Apache-2.0
+"""Generate multiple microvm configurations and run tests.
+
+Implements a component that calls a specific function in the context
+of the cartesian product of all artifact sets.
+"""
+
+import os
+from framework.artifacts import ArtifactSet
+
+
+ARTIFACTS_LOCAL_ROOT = "/tmp/ci-artifacts/"
+
+
+class TestContext:
+    """Define a context for running test fns by TestMatrix."""
+
+    __test__ = False
+
+    def __init__(self):
+        """Initialize the test context."""
+        self._context = {}
+        # User defined context.
+        self._custom = None
+
+    def set_any(self, key, value):
+        """Set any key value in the context."""
+        self._context[key] = value
+
+    @property
+    def kernel(self):
+        """Return the kernel artifact."""
+        return self._context.get('kernel', None)
+
+    @kernel.setter
+    def kernel(self, kernel):
+        """Setter for kernel artifact."""
+        self._context['kernel'] = kernel
+
+    @property
+    def disk(self):
+        """Return the disk artifact."""
+        return self._context.get('disk', None)
+
+    @disk.setter
+    def disk(self, disk):
+        """Setter for disk artifact."""
+        self._context['disk'] = disk
+
+    @property
+    def microvm(self):
+        """Return the microvm artifact."""
+        return self._context.get('microvm', None)
+
+    @microvm.setter
+    def microvm(self, microvm):
+        """Setter for kernel artifact."""
+        self._context['microvm'] = microvm
+
+    @property
+    def snapshot(self):
+        """Return the snapshot artifact."""
+        return self._context.get('snapshot', None)
+
+    @snapshot.setter
+    def snapshot(self, snapshot):
+        """Setter for snapshot artifact."""
+        self._context['snapshot'] = snapshot
+
+    @property
+    def custom(self):
+        """Return the custom context."""
+        return self._context.get('custom', None)
+
+    @custom.setter
+    def custom(self, custom):
+        """Setter for custom context."""
+        self._context['custom'] = custom
+
+
+class TestMatrix:
+    """Computes the cartesian product of artifacts."""
+
+    __test__ = False
+
+    def __init__(self,
+                 artifact_sets,
+                 context=TestContext(),
+                 cache_dir=ARTIFACTS_LOCAL_ROOT):
+        """Initialize the cache directory."""
+        self._cache_dir = cache_dir
+        self._context = context
+        # Stores the artifact sets array.
+        self._sets = artifact_sets
+        # ArtifactSet stack pointer.
+        self._set_index = 0
+        if not os.path.exists(cache_dir):
+            os.mkdir(cache_dir)
+
+    @property
+    def sets(self):
+        """Return the artifact sets."""
+        return self._sets
+
+    def add_set(self, artifact_set: ArtifactSet):
```

Comment:
This also looks unused. Can we add it when it will be used? It is easier to review this way :D.

Reply:
Removed it, the constructor is enough.